src/Psalm/Report/ByIssueSeverityReport.php

```diff
@@ -0,0 +1,180 @@
+                $snippets .= '  ' . $node_data->label . ' - ' . $this->getFileReference($node_data) . "\n";
+
+                if ($this->show_snippet) {
+                    $snippet = $node_data->snippet;
+
+                    if (!$this->use_color) {
+                        $snippets .= $snippet . "\n\n";
+                    } else {
+                        $selection_start = $node_data->from - $node_data->snippet_from;
+                        $selection_length = $node_data->to - $node_data->from;
+
+                        $snippets .= substr($snippet, 0, $selection_start)
+                            . "\e[30;47m" . substr($snippet, $selection_start, $selection_length)
+                            . "\e[0m" . substr($snippet, $selection_length + $selection_start) . "\n\n";
+                    }
+                }
+            } else {
+                $snippets .= '  ' . $node_data['label'] . "\n";
+                $snippets .= '    <no known location>' . "\n\n";
+            }
+        }
+
+        return $snippets;
+    }
+
+    /**
+     * @param IssueData|DataFlowNodeData $data
+     */
+    private function getFileReference($data): string
+    {
+        $reference = $data->file_name . ':' . $data->line_from . ':' . $data->column_from;
+
+        if (!$this->use_color) {
+            return $reference;
+        }
+
+        $file_basename = basename($data->file_name);
+        $file_path = substr($data->file_name, 0, -strlen($file_basename));
+
+        $reference = $file_path
+            . "\033[1;31m"
+            . $file_basename . ':' . $data->line_from . ':' . $data->column_from
+            . "\033[0m"
+        ;
+
+        if ($this->in_ci) {
+            return $reference;
+        }
+
+        if (null === $this->link_format) {
+            // if xdebug is not enabled, use `get_cfg_var` to get the value directly from php.ini
+            $this->link_format = ini_get('xdebug.file_link_format') ?: get_cfg_var('xdebug.file_link_format')
+                ?: 'file://%f#L%l';
+        }
+
+        $link = strtr($this->link_format, ['%f' => $data->file_path, '%l' => $data->line_from]);
+        // $reference = $data->file_name . ':' . $data->line_from . ':' . $data->column_from;
+
+
+        return "\033]8;;" . $link . "\033\\" . $reference . "\033]8;;\033\\";
+    }
+
+    public function sortIssuesByLevelAndType(): void
+    {
+        usort($this->issues_data, function (IssueData $left, IssueData $right): int {
+            $leftLevel = $left->error_level;
+            $rightLevel = $right->error_level;
+
+            if ($leftLevel != $rightLevel) {
+                if ($rightLevel > 0 && $leftLevel > 0) {
+                    return $rightLevel <=> $leftLevel;
+                }
+
+                if ($rightLevel > 0) {
+                    return -1;
+                }
+
+                return $leftLevel <=> $rightLevel;
+            }
+
+            return $left->type <=> $right->type;
```

Comment:
isn't it equivalent to `return [$left->error_level, $left->type] <=> [$right->error_level, $right->type]`?

Reply:
I think you're right, I didn't realise you can use comparison operators with arrays.

Reply:
Turns out this was almost right, but not quite. complication is that negative levels sort to the top, small positive levels sort to the bottom.